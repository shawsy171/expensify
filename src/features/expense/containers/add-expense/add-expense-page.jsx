import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';

// components
import ExpenseForm from '../../components/expense-form/expense-form';

// store
import { addExpense } from './../../store/expense/actions';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <p> this is the AddExpensePage </p>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense)),
});

AddExpensePage.propTypes = {
  addExpense: PropTypes.func.isRequired,
  history: ReactRouterPropTypes.history.isRequired,
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
