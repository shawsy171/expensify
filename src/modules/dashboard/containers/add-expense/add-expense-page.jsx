import React from 'react';

const AddExpensePage = () => (
  <p> this is the AddExpensePage </p>
);

export default AddExpensePage;