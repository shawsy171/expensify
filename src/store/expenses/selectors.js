/** filters and sorts an array of expenses 
 * @param { Array<Object> } expenses 
 * @interface Array<{
 *  text: string, 
 *  sortBy: string,
 *  startDate: number,
 *  endDate: number,
 * }>
 * @example [{
 *  description = 'Rent', 
 *  note = 'to be paid',
 *  amount = 5000,
 *  createdAt = 19299303
 * }, {
 *  description = 'Coffee', 
 *  note = 'I love coffee in the morning',
 *  amount = 300,
 *  createdAt = 85849307
 * }]
 * @param { Object } filterObject
 * @interface {
 *  text?: string, 
 *  sortBy: string,
 *  startDate?: number,
 *  endDate?: number,
 * }
 * @example {
 *  text: 'rent', 
 *  sortBy: 'amount',
 *  startDate: 100,
 *  endDate: 1250,
 * }
 * @returns { Array<Object> } expenses
 */
export const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
}