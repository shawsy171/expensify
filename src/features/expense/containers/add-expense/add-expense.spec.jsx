import React from 'react';
import { shallow } from 'enzyme';
import { historyMock } from './../../../../tests/router.mock';
import expenses from './../../../../tests/fixtures';

import { AddExpensePage } from './add-expense-page';

describe('Add Expense Component', () => {
  let wrapper;
  let history;
  let onSubmit;

  beforeEach(() => {
    onSubmit = jest.fn();
    history = { ...historyMock };
    wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history} />);
  });

  test('should render AddExpensePage Correctly ', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle onSubmit ', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onSubmit).toHaveBeenLastCalledWith(expenses[1]);
  });
});
