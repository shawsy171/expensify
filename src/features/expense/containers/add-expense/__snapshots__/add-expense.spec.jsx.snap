// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Expense Component should render AddExpensePage Correctly  1`] = `
<div>
  <p>
     this is the AddExpensePage 
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 0,
        "calenderFocused": false,
        "createdAt": "1970-01-01T00:00:00.000Z",
        "description": "",
        "error": false,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;
